transform/coroutines.go

```diff
@@ -763,6 +763,27 @@ func (c *coroutineLoweringPass) lowerCallReturn(caller *asyncFunc, call llvm.Val
 // lowerFuncCoro transforms an async function into a coroutine by lowering async operations to `llvm.coro` intrinsics.
 // See https://llvm.org/docs/Coroutines.html for more information on these intrinsics.
 func (c *coroutineLoweringPass) lowerFuncCoro(fn *asyncFunc) {
+	// Ensure that any alloca instructions in the entry block are at the start.
+	// Otherwise, block splitting would result in unintended behavior.
+	{
+		// Skip alloca instructions at the start of the block.
+		inst := fn.fn.FirstBasicBlock().FirstInstruction()
+		for !inst.IsAAllocaInst().IsNil() {
+			inst = llvm.NextInstruction(inst)
+		}
+
+		// Find any other alloca instructions and move them after the other allocas.
+		c.builder.SetInsertPointBefore(inst)
+		for !inst.IsNil() {
+			next := llvm.NextInstruction(inst)
+			if !inst.IsAAllocaInst().IsNil() {
+				inst.RemoveFromParentAsInstruction()
+				c.builder.Insert(inst)
+			}
+			inst = next
+		}
```

Comment:
This moves alloca instructions together, which I guess is fine, but I don't see how this addresses the issue I raised before?

> I don't think alloca instructions are always at the start of the entry block. I think it would be safer to check the entire entry block for alloca instructions, just in case some are not the first instruction.

Reply:
This moves all alloca instructions to the start of the entry block, so later the check does correctly apply to all allocas in the entry block. This also fixes the potential issues where an alloca could be moved by a later transformation into another block by `SplitBasicBlock`.